File: Algorithms/C++/0155/0155.cpp
class MinStack {
private:
    vector<int> data;
    vector<int> delta;
    int minIndex;
    int minElement;
public:
    MinStack() {
        minIndex = -1;
        minElement = INT_MAX;
    }

    void push(int x) {
        data.push_back(x);
        if (x < minElement) {
            int newMinIndex = data.size() - 1;
            delta.push_back(newMinIndex - minIndex);
            minIndex = newMinIndex;
            minElement = x;
        }
    }

    void pop() {
        int x = data.back();

        if (x == data[minIndex]) {
            minIndex -= delta.back();
            minElement = data[minIndex];
            delta.pop_back();
        }

        data.pop_back();
    }

    int top() {
        return data.back();
    }

    int getMin() {
        return data[minIndex];
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(x);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: Algorithms/Python/0155/0155.py
class MinStack:


    def __init__(self):
        """
        initialize your data structure here.
        """
        self.stack = []
        self.minStack = [math.inf]

    def push(self, x: int) -> None:
        self.minStack.append(min(minStack[-1], x))
        self.stack.append(x)

    def pop(self) -> None:
        self.stack.pop()
        self.minStack.pop()

    def top(self) -> int:
        return stack[-1]

    def getMin(self) -> int:
        return MinStack[-1]


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
